### DISCUSSION TITLE
Trouble With websockets dependency?

### DISCUSSION THREAD

**Discussion by pprivon** [areed1192/td-ameritrade-python-api#183] [Q&A] [answered]
Hello,

I was following along the YouTube videos on using this API and ran into a snag when I got to the Streaming API portion. I'm running python 3.9.0 on my macOS Big Sur 11.3. I installed using python -m pip install td-ameritrade-python-api so it would install all needed dependencies with my python 3.9.0 framework. 

The code I am running (comments removed to avoid comments turning into markdown headers) is:

```
import pprint

from config import TD_ACCOUNT, CLIENT_ID, REDIRECT_URI, CREDENTIALS_PATH

from td.client import TDClient

td_client = TDClient(
    client_id = CLIENT_ID, 
    redirect_uri = REDIRECT_URI, 
    credentials_path = CREDENTIALS_PATH
)

td_client.login()

td_stream_client = td_client.create_streaming_session()

td_stream_client.write_behavior(
    write='csv',
    file_path = 'stream_data.csv',
    append_mode = True
)

td_stream_client.level_one_quotes(symbols=['MSFT','AAPL'], fields=list(range(0,38)))

td_stream_client.stream()
```

The result I'm getting is:

Traceback (most recent call last):
```
  File "<FILE LOCATION>/using_td_streaming_api.py", line 31, in <module>
    td_stream_client.stream()
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/td/stream.py", line 534, in stream
    self.loop.run_until_complete(self._connect())
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/td/stream.py", line 631, in _connect
    self.connection = await websockets.client.connect(self.websocket_url)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/websockets/imports.py", line 96, in __getattr__
    raise AttributeError(f"module {package!r} has no attribute {name!r}")
AttributeError: module 'websockets' has no attribute 'client'
```

I've confirmed that the websockets/client.py file exists. Digging around on google/stack overflow they said this could be from needing websocket-client installed but when I installed that it still didn't resolve the issue. Tried uninstalling websockets and websocket-client and just having one or the other installed, tried having both installed, and none of this resolved the issue. 

Has anyone else run into this or have suggestions? 

Thanks in advance!
-Peter

EDIT:

I thought perhaps this was due to an environment issue so I went ahead and installed Anaconda. I then followed these steps:

1. I created a brand new environment: `conda create --name TradingBotEnv`
2. Activated new environment: `conda activate TradingBotEnv`
3. Installed pip: `conda install pip`
4. Installed td-ameritrade-python-api using pip: `python -m pip install td-ameritrade-python-api`
5. Ran my python script that just creates a simple GTC LIMIT order using the td-ameritrade-python-api (not streaming) and confirmed the order created successfully.
6. Ran the code I showed above to get a simple quote return using the streaming client and received the same message as above but with the new environment path created in step 1:
```
Traceback (most recent call last):
  File "<PATH TO PYTHON SCRIPT>/using_td_streaming_api.py", line 31, in <module>
    td_stream_client.stream()
  File "~/opt/anaconda3/envs/TradingBotEnv/lib/python3.9/site-packages/td/stream.py", line 534, in stream
    self.loop.run_until_complete(self._connect())
  File "~/opt/anaconda3/envs/TradingBotEnv/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "~/opt/anaconda3/envs/TradingBotEnv/lib/python3.9/site-packages/td/stream.py", line 631, in _connect
    self.connection = await websockets.client.connect(self.websocket_url)
  File "~/opt/anaconda3/envs/TradingBotEnv/lib/python3.9/site-packages/websockets/imports.py", line 96, in __getattr__
    raise AttributeError(f"module {package!r} has no attribute {name!r}")
AttributeError: module 'websockets' has no attribute 'client'
```

**Comment by pprivon** [comment, marked as answer]
I found a workaround on this but it doesn't seem like it should be needed. Granted I'm pretty new overall to python. 

I opened the td/stream.py file and right beneath:
`import websockets`

I put in the following line of code:
`from websockets import client`

I ran the script I created and successfully received my streaming quotes.

**Reply by jeffreydiehl** [reply]
Thanks for the workaround.

Haven't poked around enough to determine specifics, but I can add that the original code does not throw this error under websockets 8.1 but fails under websockets 9.1.
